### PR TITLE
Enqueue senders in Channel#close

### LINKED ISSUE #5875: Channel#close doesn't invoke senders
```
ch = Channel(Nil).new
spawn(name: "Sender") do
  loop do
    ch.send nil
  end
rescue Channel::ClosedError
  puts "Sender closed"
end

spawn(name: "Receiver") do
  loop do
    ch.receive
    Fiber.yield
  end
rescue Channel::ClosedError
  puts "Receiver closed"
end

spawn(name: "Closer") do
  sleep 0.1
  puts "Closing channel"
  ch.close
end

sleep
```

Produces: 

```
$ crystal run test/close_channel.cr 
Closing channel
Receiver closed
```
Notably doesn't output `Sender closed` as I would've expected. 

I wild guess is that Channel#close (https://github.com/crystal-lang/crystal/blob/80a975cf0a98a4ce8464d0069145131667becd70/src/concurrent/channel.cr#L31L36) has to include

```
Scheduler.enqueue @senders
@senders.clear
```

```
$ crystal -v
Crystal 0.24.2 (2018-03-10)
LLVM: 5.0.1
Default target: x86_64-apple-macosx
```

### PR DESCRIPTION
**Pull request by carlhoerberg**
Fixes #5875

### CODE REVIEW

**Comment by kostya** [comment]
any spec?

**Comment by carlhoerberg** [comment]
coming

File: src/concurrent/channel.cr [hunk 1]
```diff
@@ -269,4 +271,11 @@ class Channel::Unbuffered(T) < Channel(T)
   def full?
     @has_value || @receivers.empty?
   end
+
+  def close
+    super
+    if sender = @sender
```
Comment (RX14) [inline src/concurrent/channel.cr]: You should set `@sender = nil` here to prevent any possibility of `@sender` being enqueued twice.

Comment (carlhoerberg) [reply src/concurrent/channel.cr]: thanks!

File: src/concurrent/channel.cr [hunk 2]
```diff
@@ -269,4 +271,11 @@ class Channel::Unbuffered(T) < Channel(T)
...
+      Scheduler.enqueue sender
```
Comment (ysbaddaden) [inline src/concurrent/channel.cr]: I'm not sure but `@sender` may have been be already enqueued, see https://github.com/carlhoerberg/crystal/blob/c901b8096fff57893c46555401a395ca1c0a1aef/src/concurrent/channel.cr#L263. Maybe `@sender` should be set to `nil` in `#receive_impl`?

```crystal
@value.tap do
  @has_value = false
  Scheduler.enqueue @sender.not_nil!
  @sender = nil
end
```

**Comment by carlhoerberg** [comment]
One of the specs throws: 
```
Unhandled exception in spawn:
Error writing to socket: Broken pipe (Errno)
  from src/socket.cr:74:11 in 'unbuffered_write'
  from src/io/buffered.cr:112:14 in 'write'
  from src/io.cr:492:7 in 'write_utf8'
  from src/string.cr:4189:5 in 'to_s'
  from src/io.cr:197:5 in '<<'
  from src/http/common.cr:149:9 in 'serialize_headers'
  from src/http/common.cr:143:5 in 'serialize_headers_and_string_body'
  from src/http/common.cr:118:7 in 'serialize_headers_and_body'
  from src/http/client/response.cr:67:5 in 'to_io'
  from ~proc4Proc(Nil)@spec/std/http/client/client_spec.cr:10
  from src/fiber.cr:255:3 in 'run'
  from ~proc2Proc(Fiber, (IO::FileDescriptor | Nil))@src/fiber.cr:29
```

But it doesn't fail the spec, I guess because it's thrown in another FIber than the spec is evaluating. A bug that's always been there but just now got to the surface because of the fix?

**Comment by RX14** [comment]
@carlhoerberg that failure has been around for ages in `master`. I've been meaning to debug it but it's unrelated to this PR.

File: src/concurrent/channel.cr [hunk 3]
```diff
@@ -258,7 +260,10 @@ class Channel::Unbuffered(T) < Channel(T)
 
     @value.tap do
       @has_value = false
-      Scheduler.enqueue @sender.not_nil!
+      if sender = @sender
```
Comment (RX14) [inline src/concurrent/channel.cr]: I think we should still assert `@sender.not_nil!` here.

```cr
Scheduler.enqueue @sender.not_nil!
@sender = nil
```

Comment (carlhoerberg) [reply src/concurrent/channel.cr]: ok! not a possible race-condition to set `@sender` to `nil` after enqueuing it though?

Comment (RX14) [reply src/concurrent/channel.cr]: Not yet. `Channel` isn't thread safe currently anyway (you can see clear `@senders` after we enqueue it).

**Comment by carlhoerberg** [comment]
I'm unable to write a test case that fails for Buffered Channels in master though :/ 

Do you want me to squash the commits into one?

**Comment by carlhoerberg** [comment]
Been running this patch in a larger application a couple of days now and neither buffered nor unbuffered channels get stuck anymore.

**Pull request merged**